File: packages/native-helpers/swift-helper/Sources/SwiftHelper/RpcHandler.swift
import Foundation
import ObjCExceptionCatcher

/// Flexible RPC request that can parse any method string
struct FlexibleRPCRequest: Codable {
    let id: String
    let method: String
    let params: JSONAny?
}

class IOBridge: NSObject {
    let jsonEncoder: JSONEncoder
    let jsonDecoder: JSONDecoder
    private let accessibilityService: AccessibilityService
    private let audioService: AudioService
    let dateFormatter: DateFormatter

    init(jsonEncoder: JSONEncoder, jsonDecoder: JSONDecoder) {
        self.jsonEncoder = jsonEncoder
        self.jsonDecoder = jsonDecoder
        self.accessibilityService = AccessibilityService()
        self.audioService = AudioService()  // Audio preloaded here at startup
        self.dateFormatter = DateFormatter()
        self.dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"
        super.init()
    }

    private func logToStderr(_ message: String) {
        let timestamp = dateFormatter.string(from: Date())
        let logMessage = "[\(timestamp)] \(message)\n"
        FileHandle.standardError.write(logMessage.data(using: .utf8)!)
    }

    // Handles a single RPC Request
    func handleRpcRequest(_ request: RPCRequestSchema) {
        var rpcResponse: RPCResponseSchema

        switch request.method {
        case .getAccessibilityTreeDetails:
            // Process accessibility tree requests on dedicated thread
            AccessibilityQueue.shared.async { [weak self] in
                guard let self = self else { return }
                self.handleAccessibilityTreeDetails(request)
            }
            return

        case .getAccessibilityContext:
            // Process accessibility context requests on dedicated thread (uses v2 service)
            AccessibilityQueue.shared.async { [weak self] in
                guard let self = self else { return }
                self.handleGetAccessibilityContext(id: request.id, params: request.params)
            }
            return

        case .getAccessibilityStatus:
            handleGetAccessibilityStatus(id: request.id)
            return

        case .requestAccessibilityPermission:
            handleRequestAccessibilityPermission(id: request.id)
            return

        case .pasteText:
            logToStderr("[IOBridge] Handling pasteText for ID: \(request.id)")
            guard let paramsAnyCodable = request.params else {
                let errPayload = Error(
                    code: -32602, data: nil, message: "Missing params for pasteText")
                rpcResponse = RPCResponseSchema(error: errPayload, id: request.id, result: nil)
                sendRpcResponse(rpcResponse)
                return
            }

            do {
                let paramsData = try jsonEncoder.encode(paramsAnyCodable)
                // Corrected to use generated Swift model name from models.swift
                let pasteParams = try jsonDecoder.decode(
                    PasteTextParamsSchema.self, from: paramsData)
                logToStderr("[IOBridge] Decoded pasteParams.transcript for ID: \(request.id)")

                // Call the actual paste function (to be implemented in AccessibilityService or similar)
                let success = accessibilityService.pasteText(transcript: pasteParams.transcript)

                // Corrected to use generated Swift model name from models.swift
                let resultPayload = PasteTextResultSchema(
                    message: success ? "Pasted successfully" : "Paste failed", success: success)
                let resultData = try jsonEncoder.encode(resultPayload)
                let resultAsJsonAny = try jsonDecoder.decode(JSONAny.self, from: resultData)
                rpcResponse = RPCResponseSchema(error: nil, id: request.id, result: resultAsJsonAny)

            } catch {
                logToStderr(
                    "[IOBridge] Error processing pasteText params or operation: \(error.localizedDescription) for ID: \(request.id)"
                )
                let errPayload = Error(
                    code: -32602, data: request.params,
                    message: "Invalid params or error during paste: \(error.localizedDescription)")
                rpcResponse = RPCResponseSchema(error: errPayload, id: request.id, result: nil)
            }

        case .muteSystemAudio:
            logToStderr("[IOBridge] Handling muteSystemAudio for ID: \(request.id)")

            audioService.playSound(named: "rec-start") { [weak self] in
                guard let self = self else {
                    let timestamp = DateFormatter().string(from: Date())
                    let logMessage =
                        "[\(timestamp)] [IOBridge] self is nil in playSound completion for muteSystemAudio. ID: \(request.id)\n"
                    FileHandle.standardError.write(logMessage.data(using: .utf8)!)
                    return
                }

                self.logToStderr(
                    "[IOBridge] rec-start.mp3 finished playing successfully. Proceeding to mute system audio. ID: \(request.id)"
                )
                let success = self.accessibilityService.muteSystemAudio()
                let resultPayload = MuteSystemAudioResultSchema(
                    message: success ? "Mute command sent" : "Failed to send mute command",
                    success: success)

                var responseToSend: RPCResponseSchema
                do {
                    let resultData = try self.jsonEncoder.encode(resultPayload)
                    let resultAsJsonAny = try self.jsonDecoder.decode(
                        JSONAny.self, from: resultData)
                    responseToSend = RPCResponseSchema(
                        error: nil, id: request.id, result: resultAsJsonAny)
                } catch {
                    self.logToStderr(
                        "[IOBridge] Error encoding muteSystemAudio result: \(error.localizedDescription) for ID: \(request.id)"
                    )
                    let errPayload = Error(
                        code: -32603, data: nil,
                        message: "Error encoding result: \(error.localizedDescription)")
                    responseToSend = RPCResponseSchema(
                        error: errPayload, id: request.id, result: nil)
                }
                self.sendRpcResponse(responseToSend)
            }
            return

        case .restoreSystemAudio:
            logToStderr("[IOBridge] Handling restoreSystemAudio for ID: \(request.id)")

            let success = accessibilityService.restoreSystemAudio()
            if success {  // Play sound only if restore was successful
                audioService.playSound(named: "rec-stop")
            }
            let resultPayload = RestoreSystemAudioResultSchema(
                message: success ? "Restore command sent" : "Failed to send restore command",
                success: success)

            do {
                let resultData = try jsonEncoder.encode(resultPayload)
                let resultAsJsonAny = try jsonDecoder.decode(JSONAny.self, from: resultData)
                rpcResponse = RPCResponseSchema(error: nil, id: request.id, result: resultAsJsonAny)
            } catch {
                logToStderr(
                    "[IOBridge] Error encoding pauseSystemAudio result: \(error.localizedDescription) for ID: \(request.id)"
                )
                let errPayload = Error(
                    code: -32603, data: nil,
                    message: "Error encoding result: \(error.localizedDescription)")
                rpcResponse = RPCResponseSchema(error: nil, id: request.id, result: nil)
            }

        case .setShortcuts:
            logToStderr("[IOBridge] Handling setShortcuts for ID: \(request.id)")
            guard let paramsAnyCodable = request.params else {
                let errPayload = Error(
                    code: -32602, data: nil, message: "Missing params for setShortcuts")
                rpcResponse = RPCResponseSchema(error: errPayload, id: request.id, result: nil)
                sendRpcResponse(rpcResponse)
                return
            }

            do {
                let paramsData = try jsonEncoder.encode(paramsAnyCodable)
                let shortcutsParams = try jsonDecoder.decode(
                    SetShortcutsParamsSchema.self, from: paramsData)

                // Update the ShortcutManager with the new shortcuts
                ShortcutManager.shared.setShortcuts(
                    pushToTalk: shortcutsParams.pushToTalk,
                    toggleRecording: shortcutsParams.toggleRecording
                )

                let resultPayload = SetShortcutsResultSchema(success: true)
                let resultData = try jsonEncoder.encode(resultPayload)
                let resultAsJsonAny = try jsonDecoder.decode(JSONAny.self, from: resultData)
                rpcResponse = RPCResponseSchema(error: nil, id: request.id, result: resultAsJsonAny)

            } catch {
                logToStderr(
                    "[IOBridge] Error processing setShortcuts params: \(error.localizedDescription) for ID: \(request.id)"
                )
                let errPayload = Error(
                    code: -32602, data: request.params,
                    message: "Invalid params: \(error.localizedDescription)")
                rpcResponse = RPCResponseSchema(error: errPayload, id: request.id, result: nil)
            }

        default:
            logToStderr("[IOBridge] Method not found: \(request.method) for ID: \(request.id)")
            let errPayload = Error(
                code: -32601, data: nil, message: "Method not found: \(request.method)")
            rpcResponse = RPCResponseSchema(error: errPayload, id: request.id, result: nil)
        }
        sendRpcResponse(rpcResponse)
    }

    private func sendRpcResponse(_ response: RPCResponseSchema) {
        do {
            let responseData = try jsonEncoder.encode(response)
            if let responseString = String(data: responseData, encoding: .utf8) {
                logToStderr("[Swift Biz Logic] FINAL JSON RESPONSE to stdout: \(responseString)")
                print(responseString)
                fflush(stdout)
            }
        } catch {
            logToStderr("Error encoding RpcResponse: \(error.localizedDescription)")
        }
    }

    // Main loop for processing RPC requests from stdin
    func processRpcRequests() {
        logToStderr("IOBridge: Starting RPC request processing loop.")
        while let line = readLine(strippingNewline: true) {
            guard !line.isEmpty, let data = line.data(using: .utf8) else {
                logToStderr("Warning: Received empty or non-UTF8 line on stdin.")
                continue
            }

            do {
                let rpcRequest = try jsonDecoder.decode(RPCRequestSchema.self, from: data)
                logToStderr(
                    "IOBridge: Received RPC Request ID \(rpcRequest.id), Method: \(rpcRequest.method)"
                )
                handleRpcRequest(rpcRequest)
            } catch {
                logToStderr(
                    "Error decoding RpcRequest from stdin: \(error.localizedDescription). Line: \(line)"
                )
                // Consider sending a parse error if ID can be extracted
            }
        }
        logToStderr("IOBridge: RPC request processing loop finished (stdin closed).")
    }

    // MARK: - Async RPC Handlers

    private func handleAccessibilityTreeDetails(_ request: RPCRequestSchema) {
        var accessibilityParams: GetAccessibilityTreeDetailsParamsSchema? = nil
        logToStderr("[IOBridge] Handling getAccessibilityTreeDetails for ID: \(request.id)")

        if let paramsAnyCodable = request.params {
            do {
                let paramsData = try jsonEncoder.encode(paramsAnyCodable)
                accessibilityParams = try jsonDecoder.decode(
                    GetAccessibilityTreeDetailsParamsSchema.self, from: paramsData)
                logToStderr(
                    "[IOBridge] Decoded accessibilityParams.rootID: \(accessibilityParams?.rootID ?? "nil") for ID: \(request.id)"
                )
            } catch {
                logToStderr(
                    "[IOBridge] Error decoding getAccessibilityTreeDetails params: \(error.localizedDescription)"
                )
                let errPayload = Error(
                    code: -32602, data: request.params,
                    message: "Invalid params: \(error.localizedDescription)")
                let rpcResponse = RPCResponseSchema(error: errPayload, id: request.id, result: nil)
                sendRpcResponse(rpcResponse)
                return
            }
        }

        // Fetch REAL accessibility tree data using the service
        switch ExceptionCatcher.try({
            self.accessibilityService.fetchFullAccessibilityTree(rootId: accessibilityParams?.rootID)
        }) {
        case .success(let actualTreeData):
            logToStderr("[IOBridge] Fetched actualTreeData. Is nil? \(actualTreeData == nil). For ID: \(request.id)")

            var treeAsJsonAny: JSONAny? = nil
            if let dataToEncode = actualTreeData {
                do {
                    let encodedData = try jsonEncoder.encode(dataToEncode)
                    treeAsJsonAny = try jsonDecoder.decode(JSONAny.self, from: encodedData)
                } catch {
                    logToStderr("[IOBridge] Error encoding actualTreeData: \(error.localizedDescription)")
                }
            }

            let resultPayload = GetAccessibilityTreeDetailsResultSchema(tree: treeAsJsonAny)
            var resultAsJsonAny: JSONAny? = nil
            do {
                let resultPayloadData = try jsonEncoder.encode(resultPayload)
                resultAsJsonAny = try jsonDecoder.decode(JSONAny.self, from: resultPayloadData)
            } catch {
                logToStderr("Error encoding result: \(error.localizedDescription)")
            }
            let rpcResponse = RPCResponseSchema(error: nil, id: request.id, result: resultAsJsonAny)
            sendRpcResponse(rpcResponse)

        case .exception(let exception):
            logToStderr("[IOBridge] NSException in fetchFullAccessibilityTree: \(exception.name) - \(exception.reason)")
            let exceptionData: [String: Any] = [
                "name": exception.name,
                "reason": exception.reason,
                "callStack": exception.callStack.prefix(10).joined(separator: "\n")
            ]
            var exceptionJsonAny: JSONAny? = nil
            if let jsonData = try? JSONSerialization.data(withJSONObject: exceptionData),
               let decoded = try? jsonDecoder.decode(JSONAny.self, from: jsonData) {
                exceptionJsonAny = decoded
            }
            let errPayload = Error(
                code: -32603,
                data: exceptionJsonAny,
                message: "\(exception.name): \(exception.reason)"
            )
            let rpcResponse = RPCResponseSchema(error: errPayload, id: request.id, result: nil)
            sendRpcResponse(rpcResponse)
        }
    }

    // MARK: - Accessibility Handlers (using consolidated service)

    private func handleGetAccessibilityContext(id: String, params: JSONAny?) {
        logToStderr("[IOBridge] Handling getAccessibilityContext for ID: \(id)")

        // Parse params (default editableOnly = false per spec)
        var editableOnly = false
        if let paramsAnyCodable = params {
            do {
                let paramsData = try jsonEncoder.encode(paramsAnyCodable)
                let contextParams = try jsonDecoder.decode(GetAccessibilityContextParams.self, from: paramsData)
                editableOnly = contextParams.editableOnly ?? false
            } catch {
                logToStderr("[IOBridge] Error decoding params: \(error.localizedDescription)")
            }
        }

        // Call service with exception handling
        switch ExceptionCatcher.try({
            AccessibilityContextService.getAccessibilityContext(editableOnly: editableOnly)
        }) {
        case .success(let context):
            logToStderr("[IOBridge] Retrieved context for ID: \(id)")
            let result = GetAccessibilityContextResult(context: context)
            sendResult(id: id, result: result)

        case .exception(let exception):
            logToStderr("[IOBridge] NSException in getAccessibilityContext: \(exception.name) - \(exception.reason)")
            sendError(id: id, code: -32603, message: "\(exception.name): \(exception.reason)")
        }
    }

    private func handleGetAccessibilityStatus(id: String) {
        logToStderr("[IOBridge] Handling getAccessibilityStatus for ID: \(id)")

        let result = AccessibilityContextService.getAccessibilityStatus()
        sendResult(id: id, result: result)
    }

    private func handleRequestAccessibilityPermission(id: String) {
        logToStderr("[IOBridge] Handling requestAccessibilityPermission for ID: \(id)")

        let result = AccessibilityContextService.requestAccessibilityPermission()
        sendResult(id: id, result: result)
    }

    // MARK: - Response Helpers

    private func sendResult<T: Encodable>(id: String, result: T) {
        do {
            let resultData = try jsonEncoder.encode(result)
            let resultAsJsonAny = try jsonDecoder.decode(JSONAny.self, from: resultData)
            let rpcResponse = RPCResponseSchema(error: nil, id: id, result: resultAsJsonAny)
            sendRpcResponse(rpcResponse)
        } catch {
            logToStderr("[IOBridge] Error encoding result: \(error.localizedDescription)")
            sendError(id: id, code: -32603, message: "Error encoding result: \(error.localizedDescription)")
        }
    }

    private func sendError(id: String, code: Int, message: String) {
        let errPayload = Error(code: code, data: nil, message: message)
        let rpcResponse = RPCResponseSchema(error: errPayload, id: id, result: nil)
        sendRpcResponse(rpcResponse)
    }

}

File: packages/native-helpers/windows-helper/src/RpcHandler.cs
using System;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using WindowsHelper.Models;
using WindowsHelper.Services;

namespace WindowsHelper
{
    public class RpcHandler : IDisposable
    {
        private readonly JsonSerializerOptions jsonOptions;
        private readonly AccessibilityService accessibilityService;
        private readonly AudioService audioService;
        private readonly StaThreadRunner? staRunner;
        private Action<string>? audioCompletionHandler;
        private bool disposed;

        public RpcHandler(StaThreadRunner? staRunner, ClipboardService clipboardService)
        {
            this.staRunner = staRunner;

            // Use the generated converter settings from the models
            jsonOptions = WindowsHelper.Models.Converter.Settings;

            // Create AccessibilityService with ClipboardService
            accessibilityService = new AccessibilityService(clipboardService);

            audioService = new AudioService();
            audioService.SoundPlaybackCompleted += OnSoundPlaybackCompleted;

            if (staRunner != null)
            {
                LogToStderr("RpcHandler: STA thread dispatch enabled via StaThreadRunner");
            }
        }

        public void Dispose()
        {
            if (disposed) return;
            disposed = true;
        }

        public void ProcessRpcRequests(CancellationToken cancellationToken)
        {
            LogToStderr("RpcHandler: Starting RPC request processing loop.");

            try
            {
                string? line;
                while (!cancellationToken.IsCancellationRequested && (line = Console.ReadLine()) != null)
                {
                    if (string.IsNullOrWhiteSpace(line))
                    {
                        LogToStderr("Warning: Received empty line on stdin.");
                        continue;
                    }

                    try
                    {
                        var request = JsonSerializer.Deserialize<RpcRequest>(line, jsonOptions);
                        if (request != null)
                        {
                            LogToStderr($"RpcHandler: Received RPC Request ID {request.Id}, Method: {request.Method}");
                            _ = Task.Run(() => HandleRpcRequest(request), cancellationToken);
                        }
                    }
                    catch (JsonException ex)
                    {
                        LogToStderr($"Error decoding RpcRequest from stdin: {ex.Message}. Line: {line}");
                    }
                }
            }
            catch (Exception ex)
            {
                LogToStderr($"Fatal error in RPC processing: {ex.Message}");
            }

            LogToStderr("RpcHandler: RPC request processing loop finished.");
        }

        private async void HandleRpcRequest(RpcRequest request)
        {
            RpcResponse response;

            try
            {
                switch (request.Method)
                {
                    case Method.GetAccessibilityTreeDetails:
                        response = await HandleGetAccessibilityTreeDetails(request);
                        break;

                    case Method.GetAccessibilityContext:
                        response = await HandleGetAccessibilityContext(request);
                        break;

                    case Method.PasteText:
                        response = HandlePasteText(request);
                        break;

                    case Method.MuteSystemAudio:
                        response = await HandleMuteSystemAudio(request);
                        return; // Response sent after audio playback

                    case Method.RestoreSystemAudio:
                        response = HandleRestoreSystemAudio(request);
                        break;

                    case Method.SetShortcuts:
                        response = HandleSetShortcuts(request);
                        break;

                    default:
                        LogToStderr($"Method not found: {request.Method} for ID: {request.Id}");
                        response = new RpcResponse
                        {
                            Id = request.Id.ToString(),
                            Error = new Error
                            {
                                Code = -32601,
                                Message = $"Method not found: {request.Method}"
                            }
                        };
                        break;
                }
            }
            catch (Exception ex)
            {
                LogToStderr($"Error handling request {request.Id}: {ex.Message}");
                response = new RpcResponse
                {
                    Id = request.Id.ToString(),
                    Error = new Error
                    {
                        Code = -32603,
                        Message = $"Internal error: {ex.Message}"
                    }
                };
            }

            SendRpcResponse(response);
        }

        private async Task<RpcResponse> HandleGetAccessibilityTreeDetails(RpcRequest request)
        {
            LogToStderr($"Handling getAccessibilityTreeDetails for ID: {request.Id}");

            GetAccessibilityTreeDetailsParams? parameters = null;
            if (request.Params != null)
            {
                try
                {
                    var json = JsonSerializer.Serialize(request.Params, jsonOptions);
                    parameters = JsonSerializer.Deserialize<GetAccessibilityTreeDetailsParams>(json, jsonOptions);
                }
                catch (Exception ex)
                {
                    LogToStderr($"Error decoding params: {ex.Message}");
                    return new RpcResponse
                    {
                        Id = request.Id.ToString(),
                        Error = new Error
                        {
                            Code = -32602,
                            Message = $"Invalid params: {ex.Message}",
                            Data = request.Params
                        }
                    };
                }
            }

            // Get accessibility tree on UI thread
            var tree = await Task.Run(() => accessibilityService.FetchAccessibilityTree(parameters?.RootId));

            return new RpcResponse
            {
                Id = request.Id.ToString(),
                Result = new GetAccessibilityTreeDetailsResult { Tree = tree }
            };
        }

        private async Task<RpcResponse> HandleGetAccessibilityContext(RpcRequest request)
        {
            LogToStderr($"Handling getAccessibilityContext for ID: {request.Id}");

            GetAccessibilityContextParams? parameters = null;
            if (request.Params != null)
            {
                try
                {
                    var json = JsonSerializer.Serialize(request.Params, jsonOptions);
                    parameters = JsonSerializer.Deserialize<GetAccessibilityContextParams>(json, jsonOptions);
                }
                catch (Exception ex)
                {
                    LogToStderr($"Error decoding params: {ex.Message}");
                    return new RpcResponse
                    {
                        Id = request.Id.ToString(),
                        Error = new Error
                        {
                            Code = -32602,
                            Message = $"Invalid params: {ex.Message}",
                            Data = request.Params
                        }
                    };
                }
            }

            var editableOnly = parameters?.EditableOnly ?? false;
            var context = await Task.Run(() => accessibilityService.GetAccessibilityContext(editableOnly));

            return new RpcResponse
            {
                Id = request.Id.ToString(),
                Result = new GetAccessibilityContextResult { Context = context }
            };
        }

        private RpcResponse HandlePasteText(RpcRequest request)
        {
            LogToStderr($"Handling pasteText for ID: {request.Id}");

            if (request.Params == null)
            {
                return new RpcResponse
                {
                    Id = request.Id.ToString(),
                    Error = new Error
                    {
                        Code = -32602,
                        Message = "Missing params for pasteText"
                    }
                };
            }

            try
            {
                var json = JsonSerializer.Serialize(request.Params, jsonOptions);
                var parameters = JsonSerializer.Deserialize<PasteTextParams>(json, jsonOptions);

                if (parameters != null)
                {
                    var success = accessibilityService.PasteText(parameters.Transcript, out var errorMessage);
                    return new RpcResponse
                    {
                        Id = request.Id.ToString(),
                        Result = new PasteTextResult
                        {
                            Success = success,
                            Message = success ? "Pasted successfully" : (errorMessage ?? "Paste failed")
                        }
                    };
                }
            }
            catch (Exception ex)
            {
                LogToStderr($"Error processing pasteText: {ex}");
                return new RpcResponse
                {
                    Id = request.Id.ToString(),
                    Error = new Error
                    {
                        Code = -32603,
                        Message = $"Error during paste operation: {ex.Message}",
                        Data = ex.ToString()
                    }
                };
            }

            return new RpcResponse
            {
                Id = request.Id.ToString(),
                Error = new Error
                {
                    Code = -32603,
                    Message = "Error during paste operation"
                }
            };
        }

        private async Task<RpcResponse> HandleMuteSystemAudio(RpcRequest request)
        {
            LogToStderr($"Handling muteSystemAudio for ID: {request.Id}");

            // Store the request ID for the completion handler
            var requestId = request.Id.ToString();

            audioCompletionHandler = (id) =>
            {
                LogToStderr($"rec-start.mp3 finished playing. Proceeding to mute system audio. ID: {id}");
                var success = audioService.MuteSystemAudio();
                var response = new RpcResponse
                {
                    Id = id,
                    Result = new MuteSystemAudioResult
                    {
                        Success = success,
                        Message = success ? "Mute command sent" : "Failed to send mute command"
                    }
                };
                SendRpcResponse(response);
                audioCompletionHandler = null;
            };

            // Play sound on thread pool - NAudio handles its own threading internally
            await audioService.PlaySound("rec-start", requestId);

            // Return dummy response (real response sent after audio completion)
            return new RpcResponse { Id = request.Id.ToString() };
        }

        private RpcResponse HandleRestoreSystemAudio(RpcRequest request)
        {
            LogToStderr($"Handling restoreSystemAudio for ID: {request.Id}");

            var success = audioService.RestoreSystemAudio();
            if (success)
            {
                // Play sound asynchronously - NAudio handles its own threading, don't wait
                _ = audioService.PlaySound("rec-stop", request.Id.ToString());
            }

            return new RpcResponse
            {
                Id = request.Id.ToString(),
                Result = new RestoreSystemAudioResult
                {
                    Success = success,
                    Message = success ? "Restore command sent" : "Failed to send restore command"
                }
            };
        }

        private void OnSoundPlaybackCompleted(object? sender, string requestId)
        {
            audioCompletionHandler?.Invoke(requestId);
        }

        private RpcResponse HandleSetShortcuts(RpcRequest request)
        {
            LogToStderr($"[RpcHandler] Handling setShortcuts for ID: {request.Id}");

            try
            {
                var paramsJson = JsonSerializer.Serialize(request.Params, jsonOptions);
                var setShortcutsParams = JsonSerializer.Deserialize<SetShortcutsParams>(paramsJson, jsonOptions);

                if (setShortcutsParams == null)
                {
                    return new RpcResponse
                    {
                        Id = request.Id.ToString(),
                        Error = new Error
                        {
                            Code = -32602,
                            Message = "Invalid params: could not deserialize SetShortcutsParams"
                        }
                    };
                }

                ShortcutManager.Instance.SetShortcuts(
                    setShortcutsParams.PushToTalk?.ToArray() ?? Array.Empty<string>(),
                    setShortcutsParams.ToggleRecording?.ToArray() ?? Array.Empty<string>()
                );

                return new RpcResponse
                {
                    Id = request.Id.ToString(),
                    Result = new SetShortcutsResult { Success = true }
                };
            }
            catch (Exception ex)
            {
                LogToStderr($"[RpcHandler] Error in setShortcuts: {ex.Message}");
                return new RpcResponse
                {
                    Id = request.Id.ToString(),
                    Error = new Error
                    {
                        Code = -32603,
                        Message = $"Internal error: {ex.Message}"
                    }
                };
            }
        }

        private void SendRpcResponse(RpcResponse response)
        {
            try
            {
                var json = JsonSerializer.Serialize(response, jsonOptions);
                LogToStderr($"[RpcHandler] Sending response to stdout: {json}");
                Console.WriteLine(json);
                Console.Out.Flush();
            }
            catch (Exception ex)
            {
                LogToStderr($"Error encoding RpcResponse: {ex.Message}");
            }
        }

        private void LogToStderr(string message)
        {
            var timestamp = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss.fff");
            Console.Error.WriteLine($"[{timestamp}] {message}");
            Console.Error.Flush();
        }
    }
}
